File: src/lib/backend/analytics.ts
/**
 * Клиент analytics-API. Контракт 1:1 с `backend/app/models/analytics.py`.
 */

import { backendFetch } from "@/lib/backendClient";

export interface MistakeRow {
  entity_type: "card" | "accreditation_question";
  entity_id: string;
  wrong_count: number;
  total_attempts: number;
  last_wrong_at_ms: number;
  accuracy: number;
}

export interface MistakesResponse {
  rows: MistakeRow[];
  total: number;
  period_ms: number | null;
}

export interface LeaderboardEntry {
  rank: number;
  user_id: string;
  nickname: string | null;
  total_points: number;
  max_points: number;
  completed_at_ms: number;
}

export interface LeaderboardResponse {
  case_date: string;
  rows: LeaderboardEntry[];
  total: number;
}

export interface DailyStat {
  date: string;
  correct: number;
  wrong: number;
  total: number;
  accuracy: number;
}

export interface MonthlyStatsResponse {
  year: number;
  month: number;
  days: DailyStat[];
  total_answers: number;
  total_correct: number;
  average_accuracy: number;
}

export interface MistakesQuery {
  entity_type?: "card" | "accreditation_question";
  period_days?: number;
  limit?: number;
}

export function fetchMistakes(
  query: MistakesQuery = {},
  opts?: { signal?: AbortSignal },
): Promise<MistakesResponse> {
  return backendFetch<MistakesResponse>("/api/analytics/mistakes", {
    query: { ...query },
    signal: opts?.signal,
  });
}

export function fetchDailyCaseLeaderboard(
  caseDate: string,
  limit = 50,
  opts?: { signal?: AbortSignal },
): Promise<LeaderboardResponse> {
  return backendFetch<LeaderboardResponse>("/api/analytics/leaderboard/daily-case", {
    query: { case_date: caseDate, limit },
    signal: opts?.signal,
  });
}

export function fetchMonthlyStats(
  year: number,
  month: number,
  opts?: { signal?: AbortSignal },
): Promise<MonthlyStatsResponse> {
  return backendFetch<MonthlyStatsResponse>("/api/analytics/stats/monthly", {
    query: { year, month },
    signal: opts?.signal,
  });
}

File: backend/app/models/analytics.py
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field

EntityTypeStr = Literal["card", "accreditation_question"]


class MistakeRow(BaseModel):
    entity_type: EntityTypeStr
    entity_id: str
    wrong_count: int
    total_attempts: int
    last_wrong_at_ms: int
    accuracy: float  # (total - wrong) / total


class MistakesResponse(BaseModel):
    rows: list[MistakeRow] = Field(default_factory=list)
    total: int
    period_ms: int | None = None


class LeaderboardEntry(BaseModel):
    rank: int
    user_id: str
    nickname: str | None
    total_points: int
    max_points: int
    completed_at_ms: int


class LeaderboardResponse(BaseModel):
    case_date: str
    rows: list[LeaderboardEntry] = Field(default_factory=list)
    total: int


class DailyStat(BaseModel):
    date: str  # YYYY-MM-DD
    correct: int
    wrong: int
    total: int
    accuracy: float


class MonthlyStatsResponse(BaseModel):
    year: int
    month: int
    days: list[DailyStat] = Field(default_factory=list)
    total_answers: int
    total_correct: int
    average_accuracy: float
